llvm/lib/CodeGen/WindowsSecureHotPatching.cpp

```diff
@@ -0,0 +1,599 @@
+  }
+
+  // If the type of GV cannot contain pointers, then it cannot point to
+  // other global variables. In this case, there is no need for redirects.
+  // For example, string literals do not contain pointers.
+  return TypeContainsPointers(GV->getValueType());
+}
+
+/*
+
+Rewriting references to global variables has some complexity.
+
+For ordinary instructions that reference GlobalVariables, we rewrite the
+operand of the instruction to a Load of the __ref_* variable.
+
+For constant expressions, we have to convert the constant expression (and
+transitively all constant expressions in its parent chain) to non-constant
+expressions, i.e. to a sequence of instructions.
+
+Pass 1:
+  * Enumerate all instructions in all basic blocks.
+
+  * If an instruction references a GlobalVariable (and it is not marked
+    as being ignored), then we create (if necessary) the __ref_* variable
+    for the GlobalVariable reference. However, we do not yet modify the
+    Instruction.
+
+  * If an instruction has an operand that is a ConstantExpr and the
+    ConstantExpression tree contains a reference to a GlobalVariable, then
+    we similarly create __ref_*. Similarly, we do not yet modify the Instruction
+    or the ConstantExpr tree.
+
+After Pass 1 completes, we will know whether we found any references to
+globals in this pass.  If the function does not use any globals (and most
+functions do not use any globals), then we return immediately.
+
+If a function does reference globals, then we iterate the list of globals
+used by this function and we generate Load instructions for each (unique)
+global.
+
+Next, we do another pass over all instructions:
+
+Pass 2:
+  * Re-visit the instructions that were found in Pass 1.
+
+  * If an instruction operand is a GlobalVariable, then look up the replacement
+    __ref_* global variable and the Value that came from the Load instruction
+    for it.  Replace the operand of the GlobalVariable with the Load Value.
+
+  * If an instruction operand is a ConstantExpr, then recursively examine the
+    operands of all instructions in the ConstantExpr tree.  If an operand is
+    a GlobalVariable, then replace the operand with the result of the load
+    *and* convert the ConstantExpr to a non-constant instruction.  This
+    instruction will need to be inserted into the BB of the instruction whose
+    operand is being modified, ideally immediately before the instruction
+    being modified.
+*/
+
+// Get or create a new global variable that points to the old one and whose
+// name begins with `__ref_`.
+//
+// In hot-patched images, the __ref_* variables point to global variables in
+// the original (unpatched) image. Hot-patched functions in the hot-patch
+// image use these __ref_* variables to access global variables. This ensures
+// that all code (both unpatched and patched) is using the same instances of
+// global variables.
+//
+// The Windows hot-patch infrastructure handles initializing these __ref_*
```

Comment:
Should this be "modifying" or "redirecting" instead of "initializing"? As in, then point to the GVs in the current image but Windows will change them to point to the base image if required.

Reply:
Changed to "modifying"